### PR TITLE
Allow Chacha20-Poly1305 to use a ivlen less than 12 bytes.

### LINKED ISSUE #20084: ChaCha20-Poly1305 fails to allow you to set a different IV length
The ChaCha20-Poly1305 man page suggests that you should be able to change the IV length from the default of 12.

However the following code will fail in the second `EVP_EncryptInit_ex` call:

```` c
   if (!EVP_EncryptInit_ex(ctx, EVP_chacha20_poly1305(), NULL, NULL, NULL))
        goto err;
    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_IVLEN, ivlen, NULL) <= 0)
        goto err;
    if (!EVP_EncryptInit_ex(ctx, NULL, NULL, key, iv))
        goto err;
````

The ctrl actually succeeds, but then attempting to use an iv of that length issues an error:

````
40F70DB59A7F0000:error:1C80006D:Provider routines:ossl_cipher_generic_initiv:invalid iv length:providers/implementations/ciphers/ciphercommon.c:651:
````
Internally, in the provider implementation, there seems to be confusion between the `nonce_len` field and the `ivlen` fields.

### PR DESCRIPTION
**Pull request by slontis**
Fixes #20084

This is a provider related bug. The generic code to test the ivlen enforces a size of 12 bytes. A special flag was added to allow the generic code to do a different check. i.e. ivlen <= 12 The code to do the zero padding already existed.

A test that does the zero padding was added to a data driven evp_test.

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by hlandau** [comment]
What's the providence of the test vector?

**Comment by slontis** [comment]
I dont know where to get test vectors for this case (which is why there were none there before).

**Comment by slontis** [comment]
I am basing the test vector off of the pre existing code non provider code. It will then at least be compatible with what we had pre 3.0

File: providers/implementations/ciphers/ciphercommon.c [hunk 1]
```diff
@@ -646,8 +646,9 @@ int ossl_cipher_generic_set_ctx_params(void *vctx, const OSSL_PARAM params[])
 int ossl_cipher_generic_initiv(PROV_CIPHER_CTX *ctx, const unsigned char *iv,
                                size_t ivlen)
 {
-    if (ivlen != ctx->ivlen
-        || ivlen > sizeof(ctx->iv)) {
+    if (ivlen > sizeof(ctx->iv)
+        || (!ctx->pad_iv && ivlen != ctx->ivlen)
+        || (ctx->pad_iv && ivlen > ctx->ivlen)) {
```
Comment (t8m) [inline providers/implementations/ciphers/ciphercommon.c:651]: I'd probably write the condition as:

```suggestion
        || (!ctx->pad_iv && ivlen != ctx->ivlen)
        || ivlen > ctx->ivlen) {
```

But the existing is OK too.

**Review by t8m — APPROVED** [review]

**Comment by hlandau** [comment]
> I am basing the test vector off of the pre existing code non provider code. It will then at least be compatible with what we had pre 3.0

Perhaps we could generate a valid test vector from another cryptography library which supports short IVs?

**Comment by romen** [comment]
> > I am basing the test vector off of the pre existing code non provider code. It will then at least be compatible with what we had pre 3.0
> 
> Perhaps we could generate a valid test vector from another cryptography library which supports short IVs?

For what is worth, maybe I can help with this.

I have written for other reasons a tool to generate more ChaCha20-Poly1305 KATs for interoperability testing.
It uses the pure-Rust implementation from the [RustCrypto: ChaCha20Poly1305][crate:chacha20poly1305] crate for the underlying crypto operations.

As [mentioned here](https://github.com/openssl/openssl/issues/20084#issuecomment-1399237190) I don't think shorter than 12 bytes nonces are standardized, and as such the Rust implementation does mandate the nonce to be exactly 12-bytes long.

If I manually zero the first 4 bytes of the 12-bytes nonce to (hex)`00000000` and set the remaining ones to (hex)`0102030405060708`, and reuse the same AAD and Plaintext of the KAT included in this PR, I do get the same KAT:

```text
Cipher = chacha20-poly1305
Key = 1c9240a5eb55d38af333888604f6b5f0473917c1402b80099dca5cbc207075c0
IV = 000000000102030405060708
AAD = f33388860000000000004e91
Tag = e0723bce23528ce6ccb10ff9627038bf
Plaintext = 496e7465726e65742d4472616674732061726520647261667420646f63756d656e74732076616c696420666f722061206d6178696d756d206f6620736978206d6f6e74687320616e64206d617920626520757064617465642c207265706c616365642c206f72206f62736f6c65746564206279206f7468657220646f63756d656e747320617420616e792074696d652e20497420697320696e617070726f70726961746520746f2075736520496e7465726e65742d447261667473206173207265666572656e6365206d6174657269616c206f7220746f2063697465207468656d206f74686572207468616e206173202fe2809c776f726b20696e2070726f67496e7465726e65742d4472616674732061726520647261667420646f63756d656e74732076616c696420666f722061206d6178696d756d206f6620736978206d6f6e74687320616e64206d617920626520757064617465642c207265706c616365642c206f72206f62736f6c65746564206279206f7468657220646f63756d656e747320617420616e792074696d652e20497420697320696e617070726f70726961746520746f2075736520496e7465726e65742d447261667473206173207265666572656e6365206d6174657269616c206f7220746f2063697465207468656d206f74686572207468616e206173202fe2809c776f726b20696e2070726f67496e7465726e65742d4472616674732061726520647261667420646f63756d656e74732076616c696420666f722061206d6178696d756d206f6620736978206d
Ciphertext = 64a0861575861af460f062c79be643bd5e805cfd345cf389f108670ac76c8cb24c6cfc18755d43eea09ee94e382d26b0bdb7b73c321b0100d4f03b7f355894cf332f830e710b97ce98c8a84abd0b948114ad176e008d33bd60f982b1ff37c8559797a06ef4f0ef61c186324e2b3506383606907b6a7c02b0f9f6157b53c867e4b9166c767b804d46a59b5216cde7a4e99040c5a40433225ee282a1b0a06c523eaf4534d7f83fa1155b0047718cbc546a0d072b04b3564eea1b422273f548271a0bb2316053fa76991955ebd63159434ecebb4e466dae5a1073a6727627097a1049e617d91d361094fa68f0ff77987130305beaba2eda04df997b714d6c6f2c299da65ba25e6a85842bf0440fd98a9a2266b061c4b3a13327c090f9a0789f58aad805275e4378a525f19232bfbfb749ede38480f405cf43ec2f1f8619ebcbc80a89e92a859c7911e674977ab17d4a7126a6b8a477358ff14a344d276ef6e504e10268ac3619fcf90c2d6c03fc2e3d1f290d9bf26c1fa1495dd8f97eec6229a55c2354e4524143551a5cc370a1c622c9390530cff21c3e1ed50c5e3daf97518ccce34156bdbd7eafab8bd417aef25c6c927301731bd319d247a1d5c3186ed10bfd9a7a24bac30e3e4503ed9204154d338b79ea276e7058e7f20f4d4fd1ac93d63f611af7b6d006c2a72add0eedc497b19cb30a198816664f0da00155f2e2d6ac61045b296d614301e0ad4983308028850dd4feffe3a8163970306e4047f5a165cb4befbc129729cd2e286e837e9b606486d402acc3dec5bf8b92387f6e486f2140

```

If it helps, I can provide these two more, generated with the same utility, one with and one without AAD:

```text
Cipher = chacha20-poly1305
Key = dead00000000000000000000000000000000000000000000000000000000c0de
IV = 000000000807060504030201
AAD = c0ffee
Tag = 976578cc725e1856707aac814f31b8ca
Plaintext = 4e656c206d657a7a6f2064656c2063616d6d696e206469206e6f7374726120766974610a6d6920726974726f7661692070657220756e612073656c7661206f73637572612c0a6368c3a9206c612064697269747461207669612065726120736d6172726974612e
Ciphertext = 879dc15b904832b8c9b0c2c435f62b679cb102bc4db906e5fe599f5622573242c339840c273eb8d3e265a03331b05dcdf268fa8690ce582f2154c6fd79b3577fa04662fae623df99303e8899dca3dc7fca33b5072519d69e140e87f4aed53deff44b678d50833e

Cipher = chacha20-poly1305
Key = dead00000000000000000000000000000000000000000000000000000000c0de
IV = 000000000807060504030201
Tag = fe99f2e356ad75c8ba16ee2e58010fb4
Plaintext = 4e656c206d657a7a6f2064656c2063616d6d696e206469206e6f7374726120766974610a6d6920726974726f7661692070657220756e612073656c7661206f73637572612c0a6368c3a9206c612064697269747461207669612065726120736d6172726974612e
Ciphertext = 879dc15b904832b8c9b0c2c435f62b679cb102bc4db906e5fe599f5622573242c339840c273eb8d3e265a03331b05dcdf268fa8690ce582f2154c6fd79b3577fa04662fae623df99303e8899dca3dc7fca33b5072519d69e140e87f4aed53deff44b678d50833e

```

[crate:chacha20poly1305]: https://crates.io/crates/chacha20poly1305

**Comment by slontis** [comment]
Putting a OTC hold on this: Do we want to support shorter iv lengths to be backward compatible, (which just left zero pads the iv to 12 bytes and is non standard). Or should it produce an error?
Maybe it should work with padding for 3.0, and error in 3.1 onwards?

**Comment by t8m** [comment]
> Maybe it should work with padding for 3.0, and error in 3.1 onwards?

IMO no. The 3.0 was the right place where to disable the short nonce support if we decided so. We disabled it inadvertently though. Either we accept that and officially document it, or we fix the code to pad the shorter nonces but then it should be done in all 3.x branches.

**Comment by mattcaswell** [comment]
> IMO no. The 3.0 was the right place where to disable the short nonce support if we decided so. We disabled it inadvertently though. Either we accept that and officially document it, or we fix the code to pad the shorter nonces but then it should be done in all 3.x branches.

Since we accidentally disabled it anyway and only just now noticed I am tempted to make it official.

**Comment by beldmit** [comment]
I agree. There is no standard requiring/specifying short nonces, so we shouldn't introduce such a standard de-facto.

**Comment by hlandau** [comment]
I also don't see any reason to support short nonces. If it was an official feature of the ChaCha20-Poly1305 RFC that is domain separated from 12-byte nonces that would be one thing, but if it's just a shortcut for zero-padding, and not officially sanctioned at that, I really see no reason to support this, since any application that wants it can just do the padding itself.

**Comment by mattcaswell** [comment]
The only real reason to do this is for backwards compatibility (it was supported in 1.1.1). But since its broken in 3.0 anyway and has been since it was released we could reasonably not fix it.

**Comment by t8m** [comment]
**OTC: A different fix is needed. In 3.x fail the CTRL setting an incorrect ivlen. There should be migration-guide and CHANGES entry. In 1.1.1 just update documentation that short IV length just left-pads it with zeros and that it won't be supported in 3.0.**

**Comment by slontis** [comment]
Closing. Replaced by #20151 and #20152

**Pull request closed without merging**